internal/security/fileprovider/provider.go

```diff
@@ -59,6 +69,124 @@ func (p *fileTokenProvider) SetConfiguration(secretConfig secretstoreclient.Secr
+			p.logger.Info(fmt.Sprintf("using policy/token defaults for service %s", serviceName))
+			servicePolicy = makeDefaultTokenPolicy(serviceName)
+			createTokenParameters = makeDefaultTokenParameters(serviceName)
+		}
+
+		if serviceConfig.CustomPolicy != nil {
+			customPolicy := serviceConfig.CustomPolicy
+			if customPolicy["path"] != nil {
+				customPaths := customPolicy["path"].(map[string]interface{})
+				if servicePolicy["path"] == nil {
+					servicePolicy["path"] = make(map[string]interface{})
+				}
+				for k, v := range customPaths {
+					(servicePolicy["path"]).(map[string]interface{})[k] = v
+				}
+			}
+		}
+
+		if serviceConfig.CustomTokenParameters != nil {
+			// Custom token parameters override the defaults
+			createTokenParameters = mergeMaps(createTokenParameters, serviceConfig.CustomTokenParameters)
+		}
+
+		// Set a meta property that consuming serices can use to automatically scope secret queries
+		createTokenParameters["meta"] = map[string]interface{}{
+			"edgex-service-name": serviceName,
+		}
+
+		// Always create a policy with this name
+		policyName := "edgex-service-" + serviceName
+
+		policyBytes, err := json.Marshal(servicePolicy)
+		if err != nil {
+			p.logger.Error(fmt.Sprintf("failed encode service policy for %s: %s", serviceName, err.Error()))
+			return err
+		}
+
+		if _, err := p.vaultClient.InstallPolicy(privilegedToken, policyName, string(policyBytes)); err != nil {
+			p.logger.Error(fmt.Sprintf("failed to install policy %s: %s", policyName, err.Error()))
+			return err
+		}
+
+		var createTokenResponse interface{}
+
+		if _, err = p.vaultClient.CreateToken(privilegedToken, createTokenParameters, &createTokenResponse); err != nil {
+			p.logger.Error(fmt.Sprintf("failed to create vault token for service %s: %s", serviceName, err.Error()))
+			return err
+		}
+
+		outputTokenDir := filepath.Join(p.tokenConfig.OutputDir, serviceName)
+		outputTokenFilename := filepath.Join(outputTokenDir, p.tokenConfig.OutputFilename)
+		if err := p.fileOpener.MkdirAll(outputTokenDir, os.FileMode(0700)); err != nil {
+			p.logger.Error(fmt.Sprintf("failed to create base directory path(s) %s: %s", outputTokenDir, err.Error()))
+			return err
+		}
+
+		p.logger.Info(fmt.Sprintf("creating token file %s", outputTokenFilename))
+		writeCloser, err := p.fileOpener.OpenFileWriter(outputTokenFilename, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, os.FileMode(0600))
+		if err != nil {
+			p.logger.Error(fmt.Sprintf("failed open token file for writing %s: %s", outputTokenFilename, err.Error()))
+			return err
+		}
+		permissionable, ok := writeCloser.(permissionable)
+		if ok {
+			if serviceConfig.FilePermissions != nil &&
+				(serviceConfig.FilePermissions).ModeOctal != nil {
+				mode, err := strconv.ParseInt(*(serviceConfig.FilePermissions).ModeOctal, 8, 32)
+				if err != nil {
+					p.logger.Error(fmt.Sprintf("invalid file mode %s: %s", *(serviceConfig.FilePermissions).ModeOctal, err.Error()))
+					return err
```

Comment:
Should writeCloser be explicitly closed before this `return`?

Reply:
File contents are not written until line 183.  defer Close() should be moved to line 156.

Reply:
defered Close() is moved to before this line.